turbopack/crates/turbo-persistence/src/interning_serde/mod.rs

```diff
@@ -1,40 +1,89 @@
 //! Exposed for usage in `turbo-tasks-backend`
 
-use std::io::Write;
+use std::io::{Read, Write};
 
 use indexmap::IndexSet;
 use rustc_hash::FxBuildHasher;
-use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde::{de::DeserializeOwned, Serialize};
 use turbo_rcstr::RcStr;
 
-#[derive(Serialize, Deserialize)]
-struct Data(Vec<u8>, IndexSet<RcStr, FxBuildHasher>);
-
-pub fn to_vec<T>(config: &pot::Config, value: &T) -> pot::Result<Vec<u8>>
+pub fn to_vec<T>(config: &pot::Config, value: &T) -> anyhow::Result<(Vec<u8>, RcStrToLocalId)>
 where
     T: Serialize,
 {
     let mut vec = Vec::new();
-    to_writer(config, value, &mut vec)?;
-    Ok(vec)
+    let ser_map = to_writer(config, value, &mut vec)?;
+    Ok((vec, ser_map))
+}
+
+#[derive(Default)]
+pub struct RcStrToLocalId(pub IndexSet<RcStr, FxBuildHasher>);
+
+#[derive(Default)]
+pub struct LocalIdToGlobalId(Vec<u32>);
+
+impl From<Vec<u32>> for LocalIdToGlobalId {
+    fn from(value: Vec<u32>) -> Self {
+        Self(value)
+    }
+}
+
+impl LocalIdToGlobalId {
+    pub fn write_to(&self, writer: &mut impl Write) -> anyhow::Result<()> {
+        let len = self.0.len() as u32;
+        writer.write_all(&len.to_le_bytes())?;
+        for id in self.0.iter() {
+            writer.write_all(&id.to_le_bytes())?;
+        }
+        Ok(())
+    }
+
+    fn read_from(reader: &mut impl Read) -> anyhow::Result<Self> {
+        let mut global_ids = Vec::new();
+
+        let mut len = [0; 4];
+        reader.read_exact(&mut len)?;
+        let len = u32::from_le_bytes(len);
+        global_ids.reserve(len as usize);
+
+        for _ in 0..len {
+            let mut id = [0; 4];
+            reader.read_exact(&mut id)?;
+            global_ids.push(u32::from_le_bytes(id));
+        }
+
+        Ok(Self(global_ids))
+    }
 }
 
-pub fn to_writer<T, W>(config: &pot::Config, value: &T, writer: W) -> pot::Result<()>
+pub fn to_writer<T, W>(config: &pot::Config, value: &T, writer: W) -> anyhow::Result<RcStrToLocalId>
 where
     T: Serialize,
     W: Write,
 {
-    let (result, ser_map) = turbo_rcstr::set_ser_map(|| config.serialize(value));
-    let value = result?;
-    let data = Data(value, ser_map);
-    config.serialize_into(&data, writer)
+    let (result, ser_map) = turbo_rcstr::set_ser_map(|| config.serialize_into(value, writer));
+    result?;
+
+    Ok(RcStrToLocalId(ser_map))
 }
 
-pub fn from_slice<T>(config: &pot::Config, slice: &[u8]) -> pot::Result<T>
+pub fn from_slice<T>(
```

Comment:
I think it would be better if `to_writer` and `from_slice` would be more consistent with each other.

So instead of `query_db`, I would pass `LocalIdToRcStr`, so we don't have to query the db will deserializing. Instead we could query it before.